### PR TITLE
[MLIR][OpenMP] Changes to function-filtering pass

### PR DESCRIPTION
**Pull request by TIFitis**
Currently, when deleting the device functions in the second stage of filtering during MLIR to LLVM translation we can end up with invalid calls to these functions. This is because of the removal of the EarlyOutliningPass which would have otherwise gotten rid of any such calls.

This patch aims to alter the function filtering pass in the following way:
	- Any host function is completely removed.
	- Call to the host function are also removed and their uses replaced with Undef values.
	- Any host function with target region code is marked to be removed during the the second stage.
	- Calls to such functions are still removed and their uses replaced with Undef values.

Co-authored-by: Sergio Afonso <sergio.afonsofumero@amd.com>

### CODE REVIEW

File: flang/lib/Optimizer/Transforms/OMPFunctionFiltering.cpp
Comment (agozillon) [inline flang/lib/Optimizer/Transforms/OMPFunctionFiltering.cpp:59]: Nit: 
containt -> contain

explicitly setting -> explicitly set (or another wording if this change doesn't accurately represent the meaning)

**Review by agozillon — APPROVED** [review]
LGTM, I've tested this locally and it does indeed fix the cases specified. However, please do wait for another reviewer to approve as well (just to have a second set of eyes)!

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:2499]: I think this code or similar should remain, because otherwise host functions containing target regions calling other host functions containing target regions will give problems. This example should work, and I think this change may have broken it:

```fortran
subroutine before(x)
  integer, intent(out) :: x
  !$omp target map(from: x)
    x = 1
  !$omp end target
end subroutine

subroutine middle()
  integer :: x
  !$omp target map(from: x)
    x = 0
  !$omp end target
  call before(x)
  call after(x)
end subroutine

subroutine after(x)
  integer, intent(out) :: x
  !$omp target map(from:x)
    x = 2
  !$omp end target
end subroutine
```

File: flang/test/Lower/OpenMP/function-filtering.f90
Comment (skatrak) [inline flang/test/Lower/OpenMP/function-filtering.f90:1]: I think the MLIR check for `target_subr` can be updated, since there is no longer an early outlining pass that changes the function name for the device:
```diff
- ! MLIR-HOST: func.func @{{.*}}target_subr(
- ! MLIR-HOST: return
- ! MLIR-DEVICE: return
+ ! MLIR-ALL: func.func @{{.*}}target_subr(
```

File: flang/test/Transforms/omp-function-filtering.mlir
Comment (skatrak) [inline flang/test/Transforms/omp-function-filtering.mlir:1]: I'd suggest adding to this test another section where the removal of host function calls and insertion of `undef` values are tested.

**Review by skatrak — CHANGES_REQUESTED** [review]
Thank you Akash for this work. My suggestions are mostly for small nits and improvements to the unit tests, except for one case that may currently not be handled correctly. If I'm wrong and it already works, I think it would still be a good idea to add a test case for it.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (TIFitis) [reply mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:2499]: In the functionFiltering, all calls to the host functions are removed regardless of whether they have a target region in them. As such the above test works as is.

Comment (skatrak) [reply mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:2499]: Ah I see, thanks for clarifying!

File: flang/test/Transforms/omp-function-filtering.mlir
Comment (TIFitis) [reply flang/test/Transforms/omp-function-filtering.mlir:1]: I've added a new section, let me know if it needs anything else.

**Comment by TIFitis** [comment]
I've updated the patch to address all the reviewer comments.

File: flang/lib/Optimizer/Transforms/OMPFunctionFiltering.cpp
Comment (TIFitis) [inline flang/lib/Optimizer/Transforms/OMPFunctionFiltering.cpp:44]: @skatrak @kiranchandramohan I had to add this change in order to create the `fir::UndefOp` when the patch is run standalone. Hope this is okay.

Comment (kiranchandramohan) [reply flang/lib/Optimizer/Transforms/OMPFunctionFiltering.cpp:44]: Would adding in https://github.com/llvm/llvm-project/blob/dbd00c3b5d8a0703e1457b5575b3400e2ebc0157/flang/include/flang/Optimizer/Transforms/Passes.td#L333 work for you?

File: flang/test/Driver/OpenMP/target-filtering.f90
Comment (banach-space) [inline flang/test/Driver/OpenMP/target-filtering.f90:1]: Is this really a driver test? This patch doesn't add any driver functionality, so there should be no driver tests, right?

Looks like all you want is to verify the lowering to LLVM, but that's different to testing the driver. Please could you move this file? I mostly want to avoid @pr-subscribers-flang-driver being tagged when no driver changes are made. Thanks!

Comment (TIFitis) [reply flang/test/Driver/OpenMP/target-filtering.f90:1]: @kiranchandramohan previously suggested `flang/test/Driver/OpenMP/` as a potential directory to keep lowering to llvm IR tests. Can you please suggest a better alternative for these tests?

File: flang/lib/Optimizer/Transforms/OMPFunctionFiltering.cpp
Comment (TIFitis) [reply flang/lib/Optimizer/Transforms/OMPFunctionFiltering.cpp]: Yes this works, thanks for the suggestion.

File: flang/test/Driver/OpenMP/target-filtering.f90
Comment (kiranchandramohan) [reply flang/test/Driver/OpenMP/target-filtering.f90:1]: The right way to do LIT tests is to test a single stage at a time. You can use the following flow as guidance.
1. For tests lowering to HLFIR + FIR + OpenMP from source use flang/test/Lower/OpenMP.
2. For tests converting from FIR + OpenMP to LLVM dialect use flang/test/Fir/convert-to-llvm-openmp-and-fir.fir
3. For tests converting form LLVM + OpenMP or core-dialect+OpenMP to LLVM + OpenMP use mlir/test/Conversion/OpenMPToLLVM/
4. For test converting from LLVM + OpenMP to LLVM IR use mlir/test/Target/LLVMIR/openmp-*

Comment (kiranchandramohan) [reply flang/test/Driver/OpenMP/target-filtering.f90:1]: If you are testing a specific pass then the tests should go to flang/test/Transforms/. If there are many OpenMP related tests you can create a sub-directory there. But note that the test here should be from FIR+OpenMP to FIR+OpenMP.

Comment (TIFitis) [reply flang/test/Driver/OpenMP/target-filtering.f90:1]: We do already have tests for the partial stages.

I felt that we should have some end-to-end tests for things that rely on multiple stages. If you feel like these are unnecessary then I can get rid of the test being added in this patch, and put up a separate patch to remove the remainder.

Comment (kiranchandramohan) [reply flang/test/Driver/OpenMP/target-filtering.f90:1]: You can write an RFC to have flang/test/llvm-ir or flang/test/end-to-end get approval and then add tests there.
You can also add end to end and execution tests in https://github.com/llvm/llvm-test-suite/tree/main/Fortran.

> If you feel like these are unnecessary

I did not say this.

File: flang/test/Transforms/omp-function-filtering.mlir
Comment (skatrak) [reply flang/test/Transforms/omp-function-filtering.mlir:1]: I think this does the trick, thanks!

File: flang/test/Driver/OpenMP/target-filtering.f90
Comment (TIFitis) [reply flang/test/Driver/OpenMP/target-filtering.f90:1]: I've put up an RFC here https://discourse.llvm.org/t/rfc-flang-new-directory-for-adding-end-to-end-tests-for-lowering-to-llvm-ir-in-flang/74872.

Shall I let the `flang/test/Driver/OpenMP/target-filtering.f90` test be there for now in the Driver? Or I can remove it and add later based on the RFC.

Comment (TIFitis) [reply flang/test/Driver/OpenMP/target-filtering.f90]: I've removed this test for now so that the patch may land. Once the RFC is settled, I'll add the test accordingly.

**Review by skatrak — APPROVED** [review]
LGTM! I'd wait for a successful pre-merge buildbot run before merging, but in my opinion it should be ready to go. Thanks for the work @TIFitis.

**Pull request merged**
